src/librustdoc/clean/inline.rs

```diff
@@ -380,23 +380,29 @@ pub(crate) fn build_impl(
     let associated_trait = tcx.impl_trait_ref(did);
 
     // Do not inline compiler-internal items unless we're a compiler-internal crate.
-    let document_compiler_internal =
-        if let Some(stab) = tcx.lookup_stability(LOCAL_CRATE.as_def_id()) {
-            stab.is_unstable() && stab.feature == sym::rustc_private
-        } else {
-            false
-        };
+    let is_directly_public = |cx: &mut DocContext<'_>, did| {
+        if !cx.cache.effective_visibilities.is_directly_public(tcx, did) {
+            return false;
+        }
+        if let Some(stab) = tcx.lookup_stability(did) && stab.is_unstable() && stab.feature == sym::rustc_private {
+            let document_compiler_internal =
```

Comment:
Why computing this variable every time? I'm a bit surprised by your changes. I would have expected something like:

```rust
|cx: &mut DocContext<'_>, def_id: DefId| {
    if !cx.cache.effective_visibilities.is_directly_public(tcx, did) {
        return false;
    }

    return let Some(stab) = tcx.lookup_stability(did) && stab.is_unstable() && stab.feature == sym::rustc_private;
}
```

Then first call it with `LOCAL_CRATE.as_def_id()` stored into `document_compiler_internal` and then used it like this in the next calls:

```rust
if !document_compiler_internal && !is_directly_public(cx, did) {
    return;
}
```

Or did I miss something?

Reply:
Hum yes but recomputing `document_compiler_internal` every time seems overkill no?